From 209b311fa9011270606e8f18bb1bb56a4493973e Mon Sep 17 00:00:00 2001
From: Kenny Kessler <kenny.kessler@motorola.com>
Date: Mon, 4 May 2015 11:07:55 -0500
Subject: [PATCH 247/959] IKSWL-13073: power: smbchrg-mmi: Switch Rate Order

Turbo chargers should not be reported as Weak.
To prevent this change the order of the IF check.
Also ensure that CDP and SDP are not reported as
Weak.

Change-Id: I6cfdf2c3870dfadefcc608ad46844c6ff84d53fb
Signed-off-by: Kenny Kessler <kenny.kessler@motorola.com>
Reviewed-on: http://gerrit.mot.com/744151
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Jason Knopsnyder <jasonk@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/power/qpnp-smbcharger-mmi.c | 23 ++++++++++++++++-------
 1 file changed, 16 insertions(+), 7 deletions(-)

diff --git a/drivers/power/qpnp-smbcharger-mmi.c b/drivers/power/qpnp-smbcharger-mmi.c
index e9dcf12dcc9..e905c2b026a 100644
--- a/drivers/power/qpnp-smbcharger-mmi.c
+++ b/drivers/power/qpnp-smbcharger-mmi.c
@@ -3275,9 +3275,10 @@ static bool smbchg_hvdcp_det_check(struct smbchg_chip *chip)
 	return false;
 }
 
-#define WEAK_CHARGER_THRESHOLD 450
+#define WEAK_CHRG_THRSH 450
 static void smbchg_rate_check(struct smbchg_chip *chip)
 {
+	union power_supply_propval prop = {0,};
 	int prev_chg_rate = chip->charger_rate;
 	char *charge_rate[] = {
 		"None", "Normal", "Weak", "Turbo"
@@ -3288,14 +3289,22 @@ static void smbchg_rate_check(struct smbchg_chip *chip)
 		return;
 	}
 
-	if (smbchg_get_aicl_level_ma(chip) < WEAK_CHARGER_THRESHOLD)
-		chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_WEAK;
-	else if (smbchg_hvdcp_det_check(chip))
-		chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_TURBO;
-	else
+	if (chip->usb_psy &&
+	    !chip->usb_psy->get_property(chip->usb_psy,
+					 POWER_SUPPLY_PROP_TYPE,
+					 &prop)) {
+		if ((prop.intval == POWER_SUPPLY_TYPE_USB_CDP) ||
+		    (prop.intval == POWER_SUPPLY_TYPE_USB))
+			chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_NORMAL;
+		else if (smbchg_hvdcp_det_check(chip))
+			chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_TURBO;
+		else if (smbchg_get_aicl_level_ma(chip) < WEAK_CHRG_THRSH)
+			chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_WEAK;
+		else
+			chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_NORMAL;
+	} else
 		chip->charger_rate = POWER_SUPPLY_CHARGE_RATE_NORMAL;
 
-
 	if (prev_chg_rate != chip->charger_rate)
 		dev_err(chip->dev, "%s Charger Detected!\n",
 			charge_rate[chip->charger_rate]);
-- 
2.11.0

